src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+from enum import Enum
+from typing import Any, Iterable, List, Mapping, Optional, Sequence, Tuple, Union
+
+import fiona
+import laspy
+import rasterio
+import shapely
+from rtree import index
+
+import tiledb
+from tiledb.cloud.utilities import Profiler
+from tiledb.cloud.utilities import as_batch
+from tiledb.cloud.utilities import batch
+from tiledb.cloud.utilities import create_log_array
+from tiledb.cloud.utilities import find
+from tiledb.cloud.utilities import get_logger_wrapper
+from tiledb.cloud.utilities import max_memory_usage
+from tiledb.cloud.utilities import run_dag
+
+fiona.drvsupport.supported_drivers["TileDB"] = "arw"
+fiona.vfs.SCHEMES["tiledb"] = "tiledb"
+
+DEFAULT_RESOURCES = {"cpu": "2", "memory": "2Gi"}
+DEFAULT_IMG_NAME = "3.9-geo"
+DEFAULT_DAG_NAME = "geo-ingestion"
+
+# Default values for ingestion parameters
+RASTER_TILE_SIZE = 1024
+POINT_CLOUD_CHUNK_SIZE = 1_000_000
+GEOMETRY_CHUNK_SIZE = 100_000
+MAX_WORKERS = 40
+BATCH_SIZE = 10
+
+
+class DatasetType(Enum):
+    POINTCLOUD = 1
+    RASTER = 2
+    GEOMETRY = 3
+
+
+@dataclass
+class BoundingBox:
+    minx: float
+    miny: float
+    maxx: float
+    maxy: float
+    minz: Optional[float] = None
+    maxz: Optional[float] = None
+
+    @property
+    def bounds(self):
+        if self.minz:
+            return (self.minx, self.miny, self.minz, self.maxx, self.maxy, self.maxz)
+        else:
+            return (self.minx, self.miny, self.maxx, self.maxy)
+
+
+@dataclass
+class GeoBlockMetadata:
+    ranges: Tuple[Tuple[int, int], Tuple[int, int]]
+    files: Tuple[os.PathLike, ...]
+
+
+@dataclass
+class GeoMetadata:
+    # common properties
+    extents: BoundingBox
+    crs: str = None
+    paths: Optional[Union[List[os.PathLike], os.PathLike]] = None
+    # raster properties
+    dtype: Optional[str] = None
+    band_count: Optional[int] = None
+    res: Optional[Tuple[float, float]] = None
+    block_metadata: Optional[Tuple[GeoBlockMetadata, ...]] = None
+    # point cloud properties
+    scales: Optional[Tuple[float, float, float]] = None
+    offsets: Optional[Tuple[float, float, float]] = None
+    # geometry properties
+    geometry_schema: Optional[dict] = None
+
+
+def get_pointcloud_metadata(
+    sources: Iterable[os.PathLike],
+    *,
+    config: Mapping[str, Any] = None,
+    verbose: bool = False,
+    id: str = "pointcloud_metadata",
+    trace: bool = False,
+    log_uri: Optional[str] = None,
+) -> GeoMetadata:
+    """Return geospatial metadata for a sequence of input point cloud data files
+
+    :param sources: iterator, paths or path to process
+    :param config: dict, configuration to pass on tiledb.VFS
+    :param verbose: bool, enable verbose logging, default is False
+    :param trace: bool, enable trace logging, default is False
+    :param log_uri: Optional[str] = None,
+    :Return: GeoMetadata, a populated GeoMetadata object
+    """
+    if len(sources) == 0:
+        raise ValueError("Input point cloud datasets required")
```

Comment:
nit: since empty iterables are false-like, this can be

```python
if not sources:
  raise ValueError(...)
```

Reply:
👍